File: src/integration/wiremock/src/jvmTest/java/community/flock/wirespec/integration/wiremock/java/StubForTest.java
package community.flock.wirespec.integration.wiremock.java;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import community.flock.wirespec.integration.jackson.java.WirespecSerialization;
import community.flock.wirespec.integration.wiremock.java.generated.endpoint.GetTodoById;
import community.flock.wirespec.integration.wiremock.java.generated.endpoint.GetTodos;
import community.flock.wirespec.integration.wiremock.java.generated.model.Error;
import community.flock.wirespec.integration.wiremock.java.generated.model.TodoDto;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;

import static community.flock.wirespec.integration.wiremock.java.WirespecWireMock.wirespec;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class StubForTest {

    private WireMockServer server;
    private final HttpClient client = HttpClient.newHttpClient();
    private final WirespecSerialization serialization = new WirespecSerialization(new ObjectMapper());

    @BeforeEach
    public void startServer() {
        server = new WireMockServer(WireMockConfiguration.wireMockConfig().dynamicPort());
        server.start();
    }

    @AfterEach
    public void stopServer() {
        server.stop();
    }

    @Test
    public void stubsStaticPathEndpoint() throws Exception {
        List<TodoDto> todos = List.of(new TodoDto("1", "Buy milk", false));

        server.stubFor(wirespec(new GetTodos.Handler.Handlers()).willReturn(new GetTodos.Response200(todos), serialization));

        HttpResponse<String> response = get("/api/todos");

        assertEquals(200, response.statusCode());
        assertEquals("[{\"id\":\"1\",\"name\":\"Buy milk\",\"done\":false}]", response.body());
    }

    @Test
    public void stubsTemplatedPathEndpoint() throws Exception {
        TodoDto todo = new TodoDto("abc", "Walk dog", true);

        server.stubFor(wirespec(new GetTodoById.Handler.Handlers()).willReturn(new GetTodoById.Response200(todo), serialization));

        HttpResponse<String> response = get("/api/todos/abc");

        assertEquals(200, response.statusCode());
        assertEquals("{\"id\":\"abc\",\"name\":\"Walk dog\",\"done\":true}", response.body());
    }

    @Test
    public void stubsNon2xxResponse() throws Exception {
        Error err = new Error(404L, "not found");

        server.stubFor(wirespec(new GetTodoById.Handler.Handlers()).willReturn(new GetTodoById.Response404(err), serialization));

        HttpResponse<String> response = get("/api/todos/anything");

        assertEquals(404, response.statusCode());
        assertTrue(response.body().contains("not found"));
    }

    private HttpResponse<String> get(String path) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(server.baseUrl() + path))
                .GET()
                .build();
        return client.send(request, HttpResponse.BodyHandlers.ofString());
    }
}

File: src/integration/wiremock/src/jvmTest/kotlin/community/flock/wirespec/integration/wiremock/kotlin/StubForTest.kt
package community.flock.wirespec.integration.wiremock.kotlin

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.tomakehurst.wiremock.WireMockServer
import com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig
import community.flock.wirespec.integration.jackson.kotlin.WirespecSerialization
import community.flock.wirespec.integration.wiremock.kotlin.generated.endpoint.GetTodoById
import community.flock.wirespec.integration.wiremock.kotlin.generated.endpoint.GetTodos
import community.flock.wirespec.integration.wiremock.kotlin.generated.model.Error
import community.flock.wirespec.integration.wiremock.kotlin.generated.model.TodoDto
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class StubForTest {

    private lateinit var server: WireMockServer
    private val client = HttpClient.newHttpClient()
    private val serialization = WirespecSerialization(ObjectMapper())

    @BeforeEach
    fun startServer() {
        server = WireMockServer(wireMockConfig().dynamicPort())
        server.start()
    }

    @AfterEach
    fun stopServer() {
        server.stop()
    }

    @Test
    fun `stubs a static-path endpoint with the typed response`() {
        val todos = listOf(TodoDto(id = "1", name = "Buy milk", done = false))

        server.stubFor(stubFor(GetTodos.Handler, GetTodos.Response200(todos), serialization))

        val response = get("/api/todos")

        assertEquals(200, response.statusCode())
        assertEquals("""[{"id":"1","name":"Buy milk","done":false}]""", response.body())
    }

    @Test
    fun `stubs a templated-path endpoint with the typed response`() {
        val todo = TodoDto(id = "abc", name = "Walk dog", done = true)

        server.stubFor(stubFor(GetTodoById.Handler, GetTodoById.Response200(todo), serialization))

        val response = get("/api/todos/abc")

        assertEquals(200, response.statusCode())
        assertEquals("""{"id":"abc","name":"Walk dog","done":true}""", response.body())
    }

    @Test
    fun `stubs a non-2xx typed response with the correct status code`() {
        val err = Error(code = 404, description = "not found")

        server.stubFor(stubFor(GetTodoById.Handler, GetTodoById.Response404(err), serialization))

        val response = get("/api/todos/anything")

        assertEquals(404, response.statusCode())
        assertTrue(response.body().contains("not found"))
    }

    @Test
    fun `urlPatternFor turns wirespec path params into non-slash matchers`() {
        val pattern = urlPatternFor("/api/todos/{id}")
        assertTrue(pattern.match("/api/todos/abc").isExactMatch)
        assertTrue(pattern.match("/api/todos/123-xyz").isExactMatch)
        assertTrue(!pattern.match("/api/todos/abc/extra").isExactMatch)
        assertTrue(!pattern.match("/api/users/abc").isExactMatch)
    }

    private fun get(path: String): HttpResponse<String> {
        val request = HttpRequest.newBuilder().uri(URI.create("${server.baseUrl()}$path")).GET().build()
        return client.send(request, HttpResponse.BodyHandlers.ofString())
    }
}
